Don't use strategic merge patch on Node.Status.Addresses

danwinship opened the pull request:
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
On a node with multiple IP addresses of the same type (eg, multiple InternalIP addresses), when the set of active addresses changed, kubelet would sometimes randomly reorder them. This could cause the node to suddenly declare a different primary node IP, causing various breakage.

This is because the kubelet sync code uses strategic merge patch to update Node.Status, but Node.Status.Addresses is incorrectly annotated with a patchStrategy that only works when there is only a single address of the same type.

Even though the annotation is wrong for the API, fixing it is an API break, particularly since changing it would result in even worse behavior when the clients and servers were different versions. (The client would generate a strategic patch according to one interpretation which the server would then apply according to a different interpretation.)

So this fixes `PatchNodeStatus()` by having it edit the automatically-generated strategic merge patch to force it to use a "replace" strategy on Status.Addressess.

**Which issue(s) this PR fixes**:
none, but the newly-added unit test demonstrates the bug. (It fails against the old code.)
(Downstream bug is https://bugzilla.redhat.com/show_bug.cgi?id=1696628 but it's mostly private)

**Does this PR introduce a user-facing change?**:
```release-note
Kubelet should now more reliably report the same primary node IP even if the set of node IPs reported by the CloudProvider changes.
```

/priority important-soon

RobertKrawitz commented:
/cc @sttts @liggitt

danwinship commented:
/test pull-kubernetes-integration

liggitt commented on cmd/kubeadm/.import-restrictions:
```diff
@@ -122,6 +122,7 @@
 				"github.com/docker/go-connections/tlsconfig",
 				"github.com/docker/go-units",
 				"github.com/docker/libnetwork/ipvs",
+				"github.com/evanphx/json-patch",
```
this adds usage to pkg/util/node, which kubeadm uses, so it transitively depends on this now

liggitt commented on pkg/util/node/node.go:
```diff
@@ -26,10 +26,10 @@ import (
 
 	"k8s.io/klog"
 
+	"github.com/evanphx/json-patch"
```
nit: separate from k8s imports

liggitt commented on pkg/util/node/node.go:
```diff
@@ -182,7 +182,7 @@ func PatchNodeStatus(c v1core.CoreV1Interface, nodeName types.NodeName, oldNode
 	}
 
-	updatedNode, err := c.Nodes().Patch(string(nodeName), types.StrategicMergePatchType, patchBytes, "status")
+	updatedNode, err := c.Nodes().Patch(string(nodeName), types.MergePatchType, patchBytes, "status")
```
the main thing I'm concerned about with this change is conditions, which also had a merge patch strategy, and is the most likely of the status fields to be written by multiple clients concurrently.

danwinship replied:
Hm... yeah.

So, we could do a strategic merge of everything except the `Addresses` field, and then (if `Addresses` changed) do a second non-strategic merge of just that?

dcbw replied:
@danwinship theres only like 2 or 3 places that Addresses should change. So perhaps we create a separate PatchNodeStatusAddresses() and use that from those 3 places instead?

squeed replied:
Interesting - it looks like SetNodeConditions doesn't *really* patch... is that OK?

thockin replied:
I think this last is the sanest approach - change as little as possible.

danwinship commented:
/assign @thockin 
/cc @khenidak

thockin reviewed (commented):
I wonder if we should fix kubelet AND make the breaking change?

We know that it's exceedingly unlikely that anyone but kubelet writes this field, right?  So is it worthwhile to fix it in API so that later we can undo the fix (after compat windows?)  I'm less convinced as I write it out.

Maybe we should also add a comment on the field that it should not be updated with strategic-merge-patch becuase ... ?

danwinship commented:
> We know that it's exceedingly unlikely that anyone but kubelet writes this field, right?

That used to be true, because if anyone else wrote it, kubelet would just overwrite their changes shortly after. But kube 1.8 added:

    +       if kl.externalCloudProvider {
    +               // We rely on the external cloud provider to supply the addresses.
    +               return nil

thockin commented:
oh yeah. Right.  Well.  I guess we can either:

a) Document this as "bug: don't merge-patch this" and reach out to everyone
we can find
b) Document this as "single value per type, deal with it" (push it onto
producers of this data)

Orthogonal to that, we can:

c) Deprecate this field and replace it with a properly keyed (or
non-merged) one.  What is the proper key?

On Mon, Jul 1, 2019 at 8:49 AM Dan Winship <notifications@github.com> wrote:

> We know that it's exceedingly unlikely that anyone but kubelet writes this
> field, right?
>
> That used to be true, because if anyone else wrote it, kubelet would just
> overwrite their changes shortly after. But kube 1.8 added:
>
> +       if kl.externalCloudProvider {
> +               // We rely on the external cloud provider to supply the addresses.
> +               return nil
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/79391?email_source=notifications&email_token=ABKWAVFVUFRWAHWVI7B64JLP5IRQBA5CNFSM4H3MURFKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODY6RVVA#issuecomment-507321044>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKWAVERTSUU5SBQ2J7E3FTP5IRQBANCNFSM4H3MURFA>
> .
>

andrewsykim commented:
Worth noting that the node controller that updates those node addresses are still in-tree https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_controller.go#L129-L190 (they get vendored by external providers) and also uses `nodeutil.PatchStatus`. Larger version skew is more common for external providers though so a v1.16 kubelet may still use a different patch strategy from the cloud controller manager.

thockin commented:
Yeah, the version skew seems particularly bad, but I don't know which parts
of the logic are done where.

Frankly, I am not even sure what the "right" key should be?  There are
clearly multiple values for a type.  Is patch logic going to DTRT and
remove entries if the key were value (eg if  ode IP changed, would old
value get removed)?

On Mon, Jul 1, 2019 at 9:38 AM Andrew Sy Kim <notifications@github.com>
wrote:

> Worth noting that the node controller that updates those node addresses
> are still in-tree
> https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_controller.go#L129-L190
> (they get vendored by external providers) and also uses
> nodeutil.PatchStatus. Larger version skew is more common for external
> providers though so a v1.16 kubelet may still use a different patch
> strategy from the cloud controller manager.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/79391?email_source=notifications&email_token=ABKWAVDNDMOBLEHRQBCV6WLP5IXHBA5CNFSM4H3MURFKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODY6WBZY#issuecomment-507338983>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKWAVAQQVOGZTQXT5ORLMTP5IXHBANCNFSM4H3MURFA>
> .
>

danwinship committed:
Hack PatchNodeStatus() to override the patch type on Status.Addresses

danwinship commented:
> b) Document this as "single value per type, deal with it" (push it onto producers of this data)

We can't do that, as discussed in the SIG meeting; that would mean you couldn't have both an IPv4 and an IPv6 address.

> c) Deprecate this field and replace it with a properly keyed (or non-merged) one. What is the proper key?

With the current layout, it shouldn't be merged: it's an un-ordered array.


So, poking around the strategic merge code a bit more, I've found another answer: the client can override the server's patch-applying behavior by adding a `{"$patch": "replace"}` object to the right place in the strategic merge patch. The new patch I just pushed does that.

It's a bit ugly as-is (manually editing the patch after the strategic merge code genrates it), but we could automate it by doing something like adding a new `patchStrategyOverride` annotation:

    -    Addresses []NodeAddress `... patchStrategy:"merge" patchMergeKey:"type" ...`
    +    Addresses []NodeAddress `... patchStrategy:"merge" patchMergeKey:"type" patchStrategyOverride:"replace" ...`

This would tell the client/patch-generating side to use the specified override strategy when generating the patch, and include an explicit `$patch` directive to force the server/patch-applying side to use that strategy as well. (The server would not change its behavior based on the `patchStrategyOverride` annotation; its behavior would only be changed because it received a different patch from the client). So old clients would continue to treat the field as `merge`, and generate patches based on that, and both old and new clients would apply those patches using `merge` semantics. But new clients would treat the field as `replace`, and generate patches based on that, which explicitly include an override indicating that they're using `replace` semantics, and so both old and new clients would apply the patch using `replace` semantics. So old clients continue to get the old behavior against any server, but new clients get the improved behavior against any server.

danwinship commented:
...or we could just do the PatchNodeStatusAddresses() thing

liggitt commented:
> or we could just do the PatchNodeStatusAddresses() thing

doubling the number of updates from nodes isn't a great idea. if we can force strategic merge patch to let us set the whole list (when node clearly expects to be the only writer of that field), I think we should do that

thockin commented:
>> b) Document this as "single value per type, deal with it" (push it onto producers of this data)
>
> We can't do that, as discussed in the SIG meeting; that would mean you couldn't have both an IPv4 and an IPv6 address.

We could add a new type for "second-family internal" or somethign.  It's a hack but it covers dual-stack.  Does NOT cover multiple in same family.

I spoke to @apelisse and @lavalamp today.  They agreed that the best approach would be to just use a patch that replaces the whole list.  I don't know that a new struct tag would be well received.

We do have to consider cloud-controller-manager rollouts and version skew, too.  I'm not sure it's a problem in practice, but we should think that thru more.

kfox1111 commented:
> I wonder if we should fix kubelet AND make the breaking change?

Fix kubelet to ignore it, and in a few releases fix it with a breaking change as nothing supported would use it?

> We know that it's exceedingly unlikely that anyone but kubelet writes this field, right? So is it worthwhile to fix it in API so that later we can undo the fix (after compat windows?) I'm less convinced as I write it out.

kubectl can't change it. I tried to fix another issue. deleting the whole node object was much easier then fixing the wrong fields. :(

danwinship commented:
/remove-area kubeadm
/remove-kind api-change
/remove-sig cluster-lifecycle
/remove-sig storage

danwinship commented:
OK, updated initial comment to reflect the current state of the PR (using a single strategic patch to update NodeStatus, but with a slightly-hacked-up patch to force replace semantics on Addresses). Tests pass.

I had added a comment to api/core/v1/types.go in a previous version of the patch, but removed it because I wasn't sure exactly what we want to say?

> We do have to consider cloud-controller-manager rollouts and version skew, too. I'm not sure it's a problem in practice, but we should think that thru more.

`pkg/controller/cloud/node_controller.go` also uses `nodeutil.PatchNodeStatus()` to update the addresses, so it will get the fix as well. If anyone is trying to update addresses from out-of-tree they'd want to copy something like the `PatchNodeStatus` change into their tree.

I don't think there are any "version skew" concerns: if kubelet is managing the addresses, then it doesn't matter what code anyone else is using because kubelet will overwrite any changes they make anyway. If kubelet is not managing addresses then it doesn't matter what code kubelet would have been using, because it's not using it.

liggitt commented:
> I don't think there are any "version skew" concerns: if kubelet is managing the addresses, then it doesn't matter what code anyone else is using because kubelet will overwrite any changes they make anyway. If kubelet is not managing addresses then it doesn't matter what code kubelet would have been using, because it's not using it.

Agreed.

thockin commented:
The version skew I am thinking about is this:

Cluster is rolling out cloud-controller-manager.  CCM is running the
controller that updates node addresses with this patch.  Meanwhile, at
least one kubelet is back-rev, writing node addresses with old code.  It's
going to flap, unless CCM's loop only chooses to control nodes with kubelet
greater than version X?

@cheftako

On Tue, Jul 2, 2019 at 7:06 AM Dan Winship <notifications@github.com> wrote:

> OK, updated initial comment to reflect the current state of the PR (using
> a single strategic patch to update NodeStatus, but with a
> slightly-hacked-up patch to force replace semantics on Addresses). Tests
> pass.
>
> I had added a comment to api/core/v1/types.go in a previous version of the
> patch, but removed it because I wasn't sure exactly what we want to say?
>
> We do have to consider cloud-controller-manager rollouts and version skew,
> too. I'm not sure it's a problem in practice, but we should think that thru
> more.
>
> pkg/controller/cloud/node_controller.go also uses
> nodeutil.PatchNodeStatus() to update the addresses, so it will get the
> fix as well. If anyone is trying to update addresses from out-of-tree
> they'd want to copy something like the PatchNodeStatus change into their
> tree.
>
> I don't think there are any "version skew" concerns: if kubelet is
> managing the addresses, then it doesn't matter what code anyone else is
> using because kubelet will overwrite any changes they make anyway. If
> kubelet is not managing addresses then it doesn't matter what code kubelet
> would have been using, because it's not using it.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/79391?email_source=notifications&email_token=ABKWAVGYQ33V3J7G7WV3OEDP5NOHJA5CNFSM4H3MURFKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZBMJRI#issuecomment-507692229>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKWAVFLXLLG5WLKUWAHV2TP5NOHJANCNFSM4H3MURFA>
> .
>

liggitt commented:
> Cluster is rolling out cloud-controller-manager. CCM is running the controller that updates node addresses with this patch. Meanwhile, at least one kubelet is back-rev, writing node addresses with old code. It's going to flap, unless CCM's loop only chooses to control nodes with kubelet greater than version X?

CCM and kubelet do not fight over addresses as far as I know. If the cloud provider is external, the controller manages addresses. If the cloud provider is not external, the kubelet manages them. Am I missing a scenario where that is not the case?

thockin reviewed (commented):
That's a lot of JSON transforms.  I don't see a better answer.

I would still leave a comment in types.go, effectively:

"This field is declared as mergeable, but the merge key is not sufficiently unique, which can cause data corruption when it is merged.  Callers should instead use a full-replacement patch.  See http://pr.k8s.io/79391 for an example."

thockin commented:
When CCM has been rolled out but some of the nodes are not yet updated?

On Tue, Jul 2, 2019 at 8:46 AM Jordan Liggitt <notifications@github.com> wrote:
>
> Cluster is rolling out cloud-controller-manager. CCM is running the controller that updates node addresses with this patch. Meanwhile, at least one kubelet is back-rev, writing node addresses with old code. It's going to flap, unless CCM's loop only chooses to control nodes with kubelet greater than version X?
>
> CCM and kubelet do not fight over addresses as far as I know. If the cloud provider is external, the controller manages addresses. If the cloud provider is not external, the kubelet manages them. Am I missing a scenario where that is not the case?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub, or mute the thread.

liggitt commented:
> When CCM has been rolled out but some of the nodes are not yet updated?

Oh, are we transitioning all addresses to be managed by the CCM, not just for "external" providers?

liggitt commented on pkg/kubelet/kubelet_node_status_test.go:
```diff
@@ -2166,3 +2166,69 @@ func TestNodeStatusHasChanged(t *testing.T) {
+	existingNode := v1.Node{ObjectMeta: metav1.ObjectMeta{Name: testKubeletHostname}}
+	kubeClient.ReactionChain = fake.NewSimpleClientset(&v1.NodeList{Items: []v1.Node{existingNode}}).ReactionChain
+
+	testAddresses := [][]v1.NodeAddress{
```
a table test with existing/new address lists would be easier to follow

would like to see tests around:
* nil list -> populated list
* empty list -> populated list
* populated list -> nil list
* populated list -> empty list
* no change with multiple addresses of the same type
* 1 internalIP addresses -> 2 internalIP addresses
* 2 internalIP addresses -> 1 internalIP addresses
* 2 internalIP addresses -> 2 different internalIP addresses

and verify that constructing/applying the patch always results in a set of node addresses semantically equal to the new addresses

kfox1111 commented:
I'm trying to evaluate how much this might effect one of my systems.
> On a node with multiple IP addresses of the same type (eg, multiple InternalIP addresses), when the set of active addresses changed, kubelet would sometimes randomly reorder them.

What typically causes the set of active addresses to change? adding new interfaces/deleting interfaces only? simply restarting kubelet? rebooting? other?

danwinship commented:
> That's a lot of JSON transforms. I don't see a better answer.

Well... I could apply a JSON patch to the JSON patch... :open_mouth:
(That actually might not be terrible...?)

> Cluster is rolling out cloud-controller-manager. CCM is running the controller that updates node addresses with this patch. Meanwhile, at least one kubelet is back-rev, writing node addresses with old code. It's going to flap

If kubelet and CCM don't agree on what the correct list of addresses is, then there will be flapping regardless of who is using which algorithm, of course.

If kubelet and CCM *do* agree on what the correct list of addresses is, then it won't flap for long, because both sides will only try to patch the list if it's currently wrong. Once the side using the fixed algorithm manages to patch it to the correct value, both sides will agree that that value is correct and so they won't try to patch it again.

danwinship commented:
> What typically causes the set of active addresses to change? adding new interfaces/deleting interfaces only? simply restarting kubelet? rebooting? other?

If you do not have a cloud provider set, then the bug doesn't affect you. (I'll update the release note to clarify that.) If you do have a cloud provider set, then the set of active addresses is up to the cloud provider, but generally corresponds to adding/deleting interfaces, and maybe adding/deleting secondary IPs on those interfaces.

danwinship committed:
Document the problem with Node.Status.Addresses and strategic merge patch

andrewsykim commented:
> If kubelet and CCM don't agree on what the correct list of addresses is, then there will be flapping regardless of who is using which algorithm, of course.

This is part of the problem though, version skew is much more complicated for the out-of-tree cloud providers because we have no control over the vendored version of Kubernetes there. Ideally we can reach a place where it's reasonable to standardize this but at the moment it's not. One solution for this is

andrewsykim commented:
> If kubelet and CCM don't agree on what the correct list of addresses is, then there will be flapping regardless of who is using which algorithm, of course.

This is part of the problem though, we have less control of version skew for the external cloud provider case which makes it hard to reason about the impact of this. The more probable and dangerous case is that a cluster is using a version of a kubelet with this patch, and then the cluster is migrated to use CCM where it vendored an older version of kubernetes (say v1.15) where the CCM is back to using strategic merge patch. Ideally you deploy a version of CCM that is lock-step with the Kubernetes version but when the cloud provider integration is out-of-tree, it's much more common for those versions to skew.

thockin commented:
I think I agree with Dan, though - we have a case where a bug exists.  If
that bug exists in 2 components, they both need to be fixed.

On Tue, Jul 2, 2019 at 3:41 PM Andrew Sy Kim <notifications@github.com>
wrote:

> If kubelet and CCM don't agree on what the correct list of addresses is,
> then there will be flapping regardless of who is using which algorithm, of
> course.
>
> This is part of the problem though, we have less control of version skew
> for the external cloud provider case which makes it hard to reason about
> the impact of this. The more probable and dangerous case is that a cluster
> is using a version of a kubelet with this patch, and then the cluster is
> migrated to use CCM where it vendored an older version of kubernetes (say
> v1.15) where the CCM is back to using strategic merge patch. Ideally you
> deploy a version of CCM that is lock-step with the Kubernetes version but
> when the cloud provider integration is out-of-tree, it's much more common
> for those versions to skew.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/79391?email_source=notifications&email_token=ABKWAVEDRWNUJEVBYFVON7TP5PKQ3A5CNFSM4H3MURFKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZCX3LY#issuecomment-507870639>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKWAVCPICVP66MS4V5QV33P5PKQ3ANCNFSM4H3MURFA>
> .
>

kfox1111 commented:
> > What typically causes the set of active addresses to change? adding new interfaces/deleting interfaces only? simply restarting kubelet? rebooting? other?
> 
> If you do not have a cloud provider set, then the bug doesn't affect you. (I'll update the release note to clarify that.) If you do have a cloud provider set, then the set of active addresses is up to the cloud provider, but generally corresponds to adding/deleting interfaces, and maybe adding/deleting secondary IPs on those interfaces.

In the case of vsphere driver, it calls net.Interfaces and filters out any non vmware prefix. It returns them in the order net.Interfaces returns them. net.Interfaces calls interfaceTable(0), defined here:
https://golang.org/src/net/interface_linux.go
which uses netlink to grab the list. I don't think netlink provides any guarantee on what order...

So its possible simply restarting kubelet would cause the order to change. Is a reordering a problem, or is it only when the number of entries changes?

andrewsykim commented:
> I think I agree with Dan, though - we have a case where a bug exists. If that bug exists in 2 components, they both need to be fixed.

I agree with this too, but would prefer to take a pragmatic approach for the out-of-tree case. Maybe we can isolate a commit that fixes the patch strategy for only the CCM case and backport that to previous versions? Forcing every out-of-tree provider to upgrade to v1.16 Kubernetes will not be easy because of the way things are vendored now (vendoring k8s.io/kubernetes is required).

andrewsykim commented:
@kfox1111 for the vSphere case, it would only apply if there is more than 1 network interface with the matching prefix.

thockin commented:
Does CCM have this equivalent loop in a new controller?  Can we just make a
different PR for that?

On Tue, Jul 2, 2019 at 4:49 PM Andrew Sy Kim <notifications@github.com>
wrote:

> I think I agree with Dan, though - we have a case where a bug exists. If
> that bug exists in 2 components, they both need to be fixed.
>
> I agree with this too, but would prefer to take a pragmatic approach for
> the out-of-tree case. Maybe we can isolate a commit that fixes the patch
> strategy for only the CCM case and backport that to previous versions?
> Forcing every out-of-tree provider to upgrade to v1.16 Kubernetes will not
> be easy because of the way things are vendored now (vendoring
> k8s.io/kubernetes is required).
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/79391?email_source=notifications&email_token=ABKWAVCCZ6CYDFA2XEYS2MTP5PSR7A5CNFSM4H3MURFKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZC3KCA#issuecomment-507884808>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKWAVH5NEQWSW6LEZ45KWTP5PSR7ANCNFSM4H3MURFA>
> .
>

thockin reviewed (commented):
Thanks!

/lgtm
/approve

andrewsykim commented:
> Does CCM have this equivalent loop in a new controller? Can we just make a different PR for that?

It shares `nodeutil.PatchNodeStatus` function with the kubelet, but yeah good point we can duplicate the function locally _worse-case_ scenario. Thanks!

kfox1111 commented:
> @kfox1111 for the vSphere case, it would only apply if there is more than 1 network interface with the matching prefix.

Yeah. I see that on my cluster for sure. I'm just trying to understand how this bug will effect me until its fixed. Is restarting a kubelet (during a crash, say) enough to trigger node update failures due to interface reordering?

Just how frequent it happens and what triggers it, dictates how to respond to it. If its only when an interface is removed or added, this won't effect any of my production systems as I don't do that. If its on kubelet restart, then its fairly serious and I need to tread very lightly as I am not in full control of that... I'd maybe have to even patch it out of band until upstreamed in that case to  be safe.

danwinship commented:
Restarting won't trigger the bug

kfox1111 commented:
Ok. Thank you.

cheftako commented:
On Tue, Jul 2, 2019 at 8:51 AM Tim Hockin <notifications@github.com> wrote:

> When CCM has been rolled out but some of the nodes are not yet updated?
>
>
For the cloud node controller to set the address on the node, the node must
have the cloud taint.
The node is given the cloud taint when it is created by the kubelet.
The kubelet will only create with the cloud taint when the kubelet is
started with the external cloud provider.
So I believe we are ok.

NeilW commented:
I'll add the fault I've just seen in here in case it adds any light to the discussion.

The Brightbox Cloud Controller throws up the following error occasionally in the Node Controller when we have multiple external IPs mapped to a node

```
E0821 10:47:34.158656       1 node_controller.go:188] Error patching node with cloud ip addresses = [failed to patch status "{\"status\":{\"$setElementOrder/addresses\":[{\"type\":\"Hostname\"},{\"type\":\"InternalDNS\"},{\"type\":\"InternalIP\"},{\"type\":\"InternalIP\"},{\"type\":\"ExternalIP\"},{\"type\":\"ExternalDNS\"},{\"type\":\"ExternalIP\"},{\"type\":\"ExternalDNS\"}],\"addresses\":[{\"address\":\"109.107.37.240\",\"type\":\"ExternalIP\"},{\"address\":\"109.107.35.7\",\"type\":\"ExternalIP\"},{\"address\":\"cip-8fp4v.gb1.brightbox.com\",\"type\":\"ExternalDNS\"},{\"address\":\"cip-2mse0.gb1.brightbox.com\",\"type\":\"ExternalDNS\"}]}}" for node "srv-tryop": The order in patch list:
[map[address:109.107.37.240 type:ExternalIP] map[address:109.107.35.7 type:ExternalIP] map[address:cip-8fp4v.gb1.brightbox.com type:ExternalDNS] map[address:cip-2mse0.gb1.brightbox.com type:ExternalDNS]]
 doesn't match $setElementOrder list:
[map[type:Hostname] map[type:InternalDNS] map[type:InternalIP] map[type:InternalIP] map[type:ExternalIP] map[type:ExternalDNS] map[type:ExternalIP] map[type:ExternalDNS]]
]
```

RobertKrawitz commented:
@NeiW do you still see it in the latest k/k master?
